kernel/rros/factory.rs

```diff
@@ -252,7 +261,117 @@ impl RrosElement {
             pointer: 0 as *mut u8,
         })
     }
+
+    /// `remove_element_device`: Remove a element device from hash list
+    /// and unregister this device.
+    pub fn remove_element_device(&mut self) {
+        let fac = unsafe { &mut *(self.factory.locked_data().get()) };
+        // SAFETY: If self.dev is None, it will panic. But when we
+        // call this function, we make sure there is a device in element.
+        // So it is safe to unwrap.
+        let dev = self.dev.as_mut().unwrap();
+        dev.unregister();
+
+        if self.is_public() {
+            drop(self.cdev.take());
+        }
+
+        {
+            let _guard = fac.inside.hash_lock.as_ref().unwrap().lock();
+            self.hash.hash_del();
+        }
+    }
+
+    /// `is_public`: If element is public return ture
+    #[inline]
+    fn is_public(&self) -> bool {
+        self.clone_flags & RROS_CLONE_PUBLIC != 0x0
+    }
+
+    /// `__do_put_element`: Delete an element
+    pub fn __do_put_element(&mut self) {
+        extern "C" {
+            fn rust_helper_synchronize_rcu();
+        }
+        let fac = unsafe { &*(self.factory.locked_data().get()) };
+        if self.dev.is_none() {
+            fac.dispose.unwrap()(self);
+        }
+
+        /*
+         * self->minor won't be free for use until rros_destroy_element()
+         * is called from the disposal handler, so there is no risk of
+         * reusing it too early.
+         */
+        self.remove_element_device();
+
+        /*
+         * Serialize with rros_open_element().
+         */
+        unsafe {
+            rust_helper_synchronize_rcu();
+        }
+
+        fac.dispose.unwrap()(self);
+    }
+
+    fn do_put_element_work(work: *mut Work) {
+        let e =
+            unsafe { &mut *(kernel::container_of!(work, RrosElement, work) as *mut RrosElement) };
+        e.__do_put_element();
+    }
+
+    unsafe extern "C" fn do_put_element_irq(work: *mut IrqWork) {
+        let e = unsafe {
+            &mut *(kernel::container_of!(work, RrosElement, irq_work) as *mut RrosElement)
+        };
+        workqueue::init_work(&mut e.work as *mut Work, Self::do_put_element_work);
+        unsafe {
+            workqueue::queue_work_on(
+                bindings::WORK_CPU_UNBOUND as i32,
+                bindings::system_wq,
+                &mut e.work as *mut Work,
+            );
+        }
+    }
+
+    /// `do_put_element`: Delete an element
+    pub fn do_put_element(&mut self) {
+        extern "C" {
+            fn rust_helper_hard_irqs_disabled() -> bool;
+        }
+        if running_inband().is_err() || unsafe { rust_helper_hard_irqs_disabled() } {
```

Comment:
> Why `running_inband().is_err()` instead of something like `running_oob()`? It's a little confusing

I don't find the function `running_oob` in our kernel. I can implement it and use there.

Reply:
> You can use the built-in function [unlikely](https://doc.rust-lang.org/std/intrinsics/fn.unlikely.html) if this condition rare happens.

Thanks for your tips, using `core::intrinsics::unlikely` is better, it is seemly already been added in core crate at rust-for-linux, I didn't know it before.

Reply:
Yes. I mean the API of running_inband here is very strange.
```
pub fn running_inband() -> Result<usize> {
    let res = unsafe { rust_helper_running_inband() };
    if res == 1 {
        return Ok(0);
    }
    Err(Error::EINVAL)
}
```
I search for the running_inband method:
```
static __always_inline bool running_inband(void)
{
	return stage_level() == 0;
}

static __always_inline bool running_oob(void)
{
	return !running_inband();
}

```
I think we don not have to handle the error and just return true or false is ok.

Reply:
> Yes. I mean the API of running_inband here is very strange.
> 
> ```
> pub fn running_inband() -> Result<usize> {
>     let res = unsafe { rust_helper_running_inband() };
>     if res == 1 {
>         return Ok(0);
>     }
>     Err(Error::EINVAL)
> }
> ```
> 
> I search for the running_inband method:
> 
> ```
> static __always_inline bool running_inband(void)
> {
> 	return stage_level() == 0;
> }
> 
> static __always_inline bool running_oob(void)
> {
> 	return !running_inband();
> }
> ```
> 
> I think we don not have to handle the error and just return true or false is ok.

Ok, I will change the running_inband function and related code.

Reply:
Currently, there are a lot of codes that use running_inband().is_ok/err() to determine the current stage, so I think it would be more appropriate to create a new issue and modify this function later.
@ruiqurm